por/sigchain.go

```diff
@@ -218,5 +220,37 @@ func SignatureHashWithPenalty(sc *pb.SigChain) ([]byte, error) {
 		leftShiftBit += count * int(config.SigChainRecentMinerBitShift.GetValueAtHeight(height))
 	}
 
+	if config.SigChainSkipMinerBitShift.GetValueAtHeight(height) > 0 && config.SigChainSkipMinerBlocks > 0 {
+		sm, err := GetSkipMiner(sc.BlockHash)
+		if err != nil {
+			return nil, err
+		}
+
+		count := 0
+		for i := 1; i < sc.Length()-2; i++ {
+			skipped := skipCount(sc.Elems[i].Id, sc.Elems[i+1].Id, sm)
+			if skipped > config.SigChainSkipMinerMaxAllowed {
+				count += skipped - config.SigChainSkipMinerMaxAllowed
+			}
+		}
+
+		leftShiftBit += count * int(config.SigChainSkipMinerBitShift.GetValueAtHeight(height))
+	}
+
 	return sc.SignatureHash(height, leftShiftBit)
 }
+
+func skipCount(from, to []byte, sm SkipMiner) int {
+	dist := chord.Distance(from, to, config.NodeIDBytes*8)
+	fingerIdx := dist.BitLen() - 1
```

Comment:
Have we any distance check go ahead of skipCount?
e.g. bad case from == to from malicious sigchain.